File: app/payments/services/payment_service.py
import os
from typing import Any
from app.payments.providers.factory import get_provider
from app.payments.providers.modern_treasury import USBankAccount, USPaymentRailType
from app.payments.providers.rapyd import InternationalBankAccount, SupportedCountries

async def receive_fiat(user_id: str, amount: float, currency: str) -> Any:
    """
    Debit funds from the user's bank account (default: US ACH via Modern Treasury).
    """
    # Construct a test US bank account from env vars
    acct = USBankAccount(
        account_number=os.getenv("TEST_US_ACCOUNT"),
        routing_number=os.getenv("TEST_US_ROUTING"),
        account_type=os.getenv("TEST_US_ACCOUNT_TYPE", "checking")
    )
    provider = get_provider("US")
    return await provider.pull(
        amount,
        currency,
        acct,
        preferred_rail=os.getenv("DEFAULT_US_PULL_RAIL", USPaymentRailType.RTP),
    )

async def send_fiat(user_id: str, amount: float, currency: str, refund: bool = False) -> Any:
    """
    Credit funds to the user's bank (default: Canada via Rapyd), or refund if refund=True.
    """
    # Construct a test Canadian bank account from env vars
    acct = InternationalBankAccount(
        routing_number=os.getenv("TEST_CA_ROUTING"),
        account_number=os.getenv("TEST_CA_ACCOUNT"),
        country=SupportedCountries.CANADA
    )
    provider = get_provider("CA")
    return await provider.push(
        amount, currency, acct, metadata={"refund": refund}
    ) 

File: app/payments/services/payment_service.ts
/**
 * Payment service for TerraFlow.
 * 
 * This module provides basic payment operations for sending and
 * receiving fiat currency through payment providers.
 */

import { getProvider } from '../providers/factory';

/**
 * Receive fiat currency from a user's account
 * 
 * @param userId ID of the user to receive from
 * @param amount Amount to receive
 * @param currency Currency code (e.g., "USD")
 * @returns Transaction result
 */
export async function receiveFiat(userId: string, amount: number, currency: string): Promise<any> {
  // Determine the country code from the user profile or transaction data
  // For simplicity, defaulting to "US" in this implementation
  const countryCode = "US";
  
  // Get the appropriate payment provider for the country
  const provider = getProvider(countryCode);
  
  // Execute the pull operation
  // This would normally use account details from a database
  return await provider.pull(amount, currency, userId);
}

/**
 * Send fiat currency to a user's account
 * 
 * @param userId ID of the user to send to
 * @param amount Amount to send
 * @param currency Currency code (e.g., "USD")
 * @param isRefund Whether this is a refund transaction
 * @returns Transaction result
 */
export async function sendFiat(
  userId: string,
  amount: number,
  currency: string,
  isRefund: boolean = false
): Promise<any> {
  // Determine the country code from the user profile or transaction data
  // For simplicity, defaulting to "US" in this implementation
  const countryCode = "US";
  
  // Get the appropriate payment provider for the country
  const provider = getProvider(countryCode);
  
  // Execute the push operation
  // This would normally use account details from a database
  const metadata = isRefund 
    ? { refund: true, original_user_id: userId } 
    : { user_id: userId };
    
  return await provider.push(amount, currency, userId, metadata);
} 